README.rst
..
    This file is part of Invenio.
    Copyright (C) 2017 CERN.

    Invenio is free software; you can redistribute it
    and/or modify it under the terms of the GNU General Public License as
    published by the Free Software Foundation; either version 2 of the
    License, or (at your option) any later version.

    Invenio is distributed in the hope that it will be
    useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Invenio; if not, write to the
    Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
    MA 02111-1307, USA.

    In applying this license, CERN does not
    waive the privileges and immunities granted to it by virtue of its status
    as an Intergovernmental Organization or submit itself to any jurisdiction.

=========================
 Invenio-Files-Processor
=========================

.. image:: https://img.shields.io/travis/inveniosoftware/invenio-files-processor.svg
        :target: https://travis-ci.org/inveniosoftware/invenio-files-processor

.. image:: https://img.shields.io/coveralls/inveniosoftware/invenio-files-processor.svg
        :target: https://coveralls.io/r/inveniosoftware/invenio-files-processor

.. image:: https://img.shields.io/github/tag/inveniosoftware/invenio-files-processor.svg
        :target: https://github.com/inveniosoftware/invenio-files-processor/releases

.. image:: https://img.shields.io/pypi/dm/invenio-files-processor.svg
        :target: https://pypi.python.org/pypi/invenio-files-processor

.. image:: https://img.shields.io/github/license/inveniosoftware/invenio-files-processor.svg
        :target: https://github.com/inveniosoftware/invenio-files-processor/blob/master/LICENSE

Invenio-Files-Processor is a module for processing files.

*This is an experimental developer preview release.*

Further documentation is available on
https://invenio-files-processor.readthedocs.io/


RELEASE-NOTES.rst
================================
 Invenio-Files-Processor v0.1.0
================================

Invenio-Files-Processor v0.1.0 was released on TBD, 2017.

About
-----

Invenio-Files-Processor module is a module for processing files.

*This is an experimental developer preview release.*

What's new
----------

- Initial public release.

Installation
------------

   $ pip install invenio-files-processor==0.1.0

Documentation
-------------

   https://invenio-files-processor.readthedocs.io/

Happy hacking and thanks for flying Invenio-Files-Processor.

| Invenio Development Team
|   Email: info@inveniosoftware.org
|   IRC: #invenio on irc.freenode.net
|   Twitter: https://twitter.com/inveniosoftware
|   GitHub: https://github.com/inveniosoftware/invenio-files-processor
|   URL: http://inveniosoftware.org


tests/test_pdfmetadata.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2017 CERN.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, CERN does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

"""pdfmetadata tests."""

from __future__ import absolute_import, print_function

import importlib
import os
from io import open

import pytest
from pkg_resources import EntryPoint

from invenio_files_processor.processors import pdfmetadata


class MockEntryPoint(EntryPoint):
    """Mocking of entrypoint."""

    def load(self):
        """Mock load entry point."""
        return importlib.import_module(self.module_name)


def login_user(client, user):
    """Log in a specified user."""
    with client.session_transaction() as sess:
        sess['user_id'] = user.id if user else None
        sess['_fresh'] = True


def test_permission(mocker, client, users, pdf_obj):
    """Test the file permission."""
    mocker.patch('pkg_resources.iter_entry_points',
                 return_value=[
                     MockEntryPoint(
                         'pdfmetadata',
                         'invenio_files_processor.processors.pdfmetadata')
                 ],
                 auto_spec=True)

    mocker.patch('invenio_files_processor.processors.pdfmetadata.process',
                 return_value={},
                 auto_spec=True)

    login_user(client, users['read'])
    resp = client.post(
        '/filesprocessor/pdfmetadata/{}'.format(pdf_obj.version_id),
        data={
            'grobid': True
        }
    )
    assert resp.status_code == 200

    login_user(client, users['non-read'])
    resp = client.post(
        '/filesprocessor/pdfmetadata/{}'.format(pdf_obj.version_id))
    # see invenio_files_rest.views.check_permission for the description of
    # error codes (401, 403 and 404)
    assert resp.status_code == 404


def test_can_process(pdf_obj, fits_obj):
    """Test the can_process function."""
    assert pdfmetadata.can_process(pdf_obj)
    assert not pdfmetadata.can_process(fits_obj)


def test_process(mocker, pdf_obj, tei_xml):
    """Test the processing function."""
    mocker.patch(
        'invenio_files_processor.processors.pdfmetadata.process_pdf_stream',
        return_value=tei_xml, auto_spec=True)

    metadata = pdfmetadata.process(pdf_obj, {'grobid': True})
    assert metadata['title'] == "The Need to Fairly Confront Spin-1 for " \
        "the New Higgs-like Particle"
    assert metadata['description'] == (
        "Spin-1 was ruled out early in LHC "
        "reports of a new particle with mass near 125 GeV. Actually the spin-1"
        " possibility was dismissed on false premises, and remains open. "
        "Model-independent classification based on Lorentz invariance permits "
        "nearly two dozen independent amplitudes for spin-1 to two vector "
        "particles, of which two remain with on-shell photons. The Landau-Yang"
        " theorems are inadequate to eliminate spin-1. Theoretical prejudice "
        "to close the gaps is unreliable, and a fair consideration based on "
        "experiment is needed. A spin-1 field can produce the resonance "
        "structure observed in invariant mass distributions, and also produce "
        "the same angular distribution of photons and ZZ decays as spin-0. "
        "However spin-0 cannot produce the variety of distributions made by "
        "spin-1. The Higgs-like pattern of decay also cannot rule out spin-1 "
        "without more analysis. Upcoming data will add information, which "
        "should be analyzed giving spin-1 full and unbiased consideration "
        "that has not appeared before."
    )
    assert metadata['keywords'] is None
    assert metadata['creators'] == [
        {
            'affiliation': 'University of Kansas',
            'name': 'John P. Ralston'
        },
        {
            'affiliation': None,
            'name': 'AUTHOR2_FORNAME AUTHOR2_SURNAME'
        }
    ]


invenio_files_processor/views.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2017 CERN.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, CERN does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

"""Invenio module for file processing tasks."""

from __future__ import absolute_import, print_function

from flask import Blueprint, abort, current_app, jsonify, request
from invenio_files_rest.models import ObjectVersion
from invenio_files_rest.views import ObjectResource

from .proxies import current_processor

blueprint = Blueprint(
    'invenio_files_processor',
    __name__,
    url_prefix='/filesprocessor',
    template_folder='templates',
    static_folder='static',
)


@blueprint.route("/<processor_name>/<version_id>", methods=['POST'])
def extract_metadata(processor_name=None, version_id=None):
    """Main file processing endpoint."""
    processor = current_processor.get_processor(processor_name)
    object_version = ObjectVersion.query.filter_by(version_id=version_id).one()

    # this function will check for  'object-read' permission
    ObjectResource.check_object_permission(object_version)

    if processor.can_process(object_version):
        try:
            metadata = processor.process(object_version, request.get_json())
            return jsonify(metadata)
        except Exception:
            current_app.logger.warning(
                ('The processor {} fails when processing the file {}.'.format(
                    processor_name, version_id)),
                exc_info=True
            )
            abort(500)
    else:
        current_app.logger.warning(
            ('The processor {} cannot process the file {}.'.format(
                processor_name, version_id)), exc_info=True
        )
        abort(500)


invenio_files_processor/__init__.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2017 CERN.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, CERN does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

"""Invenio-Files-Processor module is a module for processing files."""

from __future__ import absolute_import, print_function

from .ext import InvenioFilesProcessor
from .version import __version__

__all__ = ('__version__', 'InvenioFilesProcessor')


invenio_files_processor/processors/pdfmetadata.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2016 CERN.
#
# Invenio is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307, USA.
#
# In applying this license, CERN does not
# waive the privileges and immunities granted to it by virtue of its status
# as an Intergovernmental Organization or submit itself to any jurisdiction.

"""Implementations of different file processors."""

from __future__ import absolute_import, print_function

from io import open

import requests
import textract
from flask import abort, current_app
from invenio_files_rest.models import FileInstance
from invenio_grobid.api import process_pdf_stream
from invenio_grobid.errors import GrobidRequestError
from invenio_grobid.mapping import tei_to_dict


def can_process(object_version):
    """Check if given file can be processed by the processor.

    :param object_version: A ObjectVersion object representing the files
    :type object_version: invenio_files_rest.models.ObjectVersion
    """
    # For now, we only check the filetype
    return object_version.mimetype == 'application/pdf'


OpenAIRE_MINING_URL = 'http://mining.openaire.eu/openaireplus/analyze'


def _extract_project_info(fulltext, setting):
    """Extract the project information using the OpenAIRE mining service.

    :param fulltext: The fulltext string
    :type fulltext: str
    :param setting: The setting of the mining service.
    :type setting: dict

    The `setting` parameter is a dictionary with two fields.

    - `datacitations`: 'on' or 'off' (default: 'off')
    - `classification`: 'on' or 'off' (default: 'off')

    See http://mining.openaire.eu/openaireplus/analyze for details.
    """
    try:
        resp = requests.post(OpenAIRE_MINING_URL, data=[
                             ('document', fulltext),
                             ('datacitations', 'on' if setting.get(
                                 'datacitations') else 'off'),
                             ('classification', 'on' if setting.get(
                                 'classification') else 'off')
                             ])
        return resp.json()
    except ConnectionError:
        current_app.logger.warning(
            'Cannot connect to OpenAIRE mining service',
            exc_info=True
        )
        abort(500)


def process(object_version, setting={}):
    """Process the file with Grobid.

    :param object_version: A ObjectVersion object representing the files
    :type object_version: invenio_files_rest.models.ObjectVersion
    :param setting: The processor configuration.
    :type setting: dict

    An example of ``setting`` is as follows.

    .. code-block:: python

       {
           'grobid': True,
           'openaire': {
               'datacitations': 'on',
               'classification': 'on'
           }
       }
    """
    file_instance = FileInstance.get(object_version.file_id)
    xml = None
    if setting.get('grobid'):
        with open(file_instance.uri, 'rb') as pdf_file:
            try:
                xml = process_pdf_stream(pdf_file)
            except GrobidRequestError:
                current_app.logger.warning(
                    'grobid request fails when processing the file {}.'.format(
                        object_version.version_id),
                    exc_info=True
                )
                abort(500)
        r = tei_to_dict(xml)

    project_info = None
    if 'openaire' in setting:
        fulltext = textract.process(file_instance.uri, extension='pdf')
        project_info = _extract_project_info(fulltext, setting['openaire'])

    # showing the JSON for debugging
    metadata = dict(
        title=r.get('title'),
        description=r.get('abstract'),
        keywords=[it['value']
                  for it in r['keywords']] if 'keywords' in r else None,
        creators=[dict(
            name=it['name'],
            affiliation=it['affiliations'][0]['value']
            if len(it['affiliations']) else None
        ) for it in r['authors']] if 'authors' in r else None,
        project_info=project_info
    )

    return metadata
